disperser/dataapi/metrics.go

```diff
@@ -236,3 +245,157 @@ func (collector *DynamoDBCollector) Collect(ch chan<- prometheus.Metric) {
+				ConstLabels: prometheus.Labels{
+					"status": status.String(),
+				},
+			},
+		)
+		prometheus.MustRegister(gauge)
+		collector.statusMetrics[status] = &BlobStatusMetrics{
+			gauge:        gauge,
+			currentValue: 0,
+		}
+	}
+
+	// Do initial count
+	collector.updateCounts(context.Background())
+
+	return collector
+}
+
+// countBlobsWithStatus counts blobs for a specific status with pagination and timeout handling
+func (collector *DynamoDBCollectorV2) countBlobsWithStatus(ctx context.Context, status commonv2.BlobStatus) (int32, error) {
+	var totalCount int32
+	var cursor *blobstorev2.StatusIndexCursor
+
+	for {
+		select {
+		case <-ctx.Done():
+			return totalCount, ctx.Err()
+		default:
+			blobs, nextCursor, err := collector.blobMetadataStore.GetBlobMetadataByStatusPaginated(ctx, status, cursor, 100)
+			if err != nil {
+				return totalCount, err
+			}
+
+			count := int32(len(blobs))
+			totalCount += count
+
+			collector.logger.Debug("Got partial count for status",
+				"status", status.String(),
+				"partial_count", count,
+				"running_total", totalCount,
+				"has_more", nextCursor != nil,
+			)
+
+			if count == 0 || nextCursor == nil {
+				return totalCount, nil
+			}
+			cursor = nextCursor
+		}
+	}
+}
+
+func (collector *DynamoDBCollectorV2) updateCounts(ctx context.Context) {
+	collector.logger.Debug("Starting blob status count update")
+	startTime := time.Now()
+
+	for status, metrics := range collector.statusMetrics {
+		statusCtx, cancel := context.WithTimeout(ctx, 5*time.Second)
+
+		totalCount, err := collector.countBlobsWithStatus(statusCtx, status)
+		defer cancel()
+
+		if err != nil {
+			if err == context.DeadlineExceeded {
+				collector.logger.Warn("Timeout while counting blobs",
+					"status", status.String(),
+					"current_count", metrics.currentValue,
+					"using_stale_data", true,
+				)
+				continue // Keep using the last known value
+			}
+			collector.logger.Error("Failed to get count of blob metadata by status",
+				"status", status,
+				"err", err,
+				"current_value", metrics.currentValue,
+			)
+			continue
+		}
+
+		metrics.gauge.Set(float64(totalCount))
+		metrics.currentValue = float64(totalCount)
+
+		collector.logger.Debug("Updated blob status count",
+			"status", status.String(),
+			"count", totalCount,
+		)
+	}
+
+	collector.logger.Debug("Completed blob status count update",
+		"duration_ms", time.Since(startTime).Milliseconds(),
+	)
+}
+
+func (collector *DynamoDBCollectorV2) Describe(ch chan<- *prometheus.Desc) {
+	for _, metrics := range collector.statusMetrics {
+		ch <- metrics.gauge.Desc()
+	}
+}
+
+func (collector *DynamoDBCollectorV2) Collect(ch chan<- prometheus.Metric) {
+	collector.logger.Debug("Prometheus scrape triggered, updating counts")
+	startTime := time.Now()
+
+	// Create a context with timeout for the entire collection
+	ctx, cancel := context.WithTimeout(context.Background(), 8*time.Second)
```

Comment:
comment the reasons for 8s choice

Reply:
I read somewhere that the default scrape timeoit for prometheus is 10s so choosing 8 to be slightly less. Will add a comment.